### PR TITLE
Locally connected layer

### PR DESCRIPTION
**Pull request by jackculpepper**
This PR implements a locally connected layer, as described in [1], for example.

It's similar to convolution, but there is no weight sharing.

Ye Lu and I worked on this together. Working code is mostly his. Bugs are mine.

We wrote this a while back, before the convolution layer supported non-square filters and strides. I can add support for that if there is interest.

Looking forward to reading your comments.

[1] Yaniv Taigman, Ming Yang, Marc'Aurelio Ranzato, Lior Wolf. DeepFace: Closing the Gap to Human-Level Performance in Face Verification. CVPR 2014. [pdf](http://www.cv-foundation.org/openaccess/content_cvpr_2014/papers/Taigman_DeepFace_Closing_the_2014_CVPR_paper.pdf)

### CODE REVIEW

**Comment by Prasanna1991** [comment]
Should one need to update the caffe to use layers like "LOCAL_WEIGHTED_CONVOLUTION" ?? And is this layer the same as the one used by DeepFace for locally connected layer as: L4, L5, L6?? Thanks in advance!!

**Comment by singetta** [comment]
@jackculpepper thank you for your projects about locally connected layer. i also find and i want to develop as described in [1].
i installed your projects but i don't know how to do use this layers.
i saw your examples 
 - examples/mnist/train_lenet_local.sh
 - examples/siamese/train_mnist_siamese_local.sh
but this examples just know about difference conv layers and locally layers.
anyway i use your locally layers and write train.prototxt for learning caffe as described in [1] and result accuracy is just 0.23...
so i don't know how to do develop this paper. if possible, please comment and upload your train.prototxt? thank you. 

this train.prototxt is below.. 

name: "FaceNet"
layers {
  name: "data"
  type: DATA
  top: "data"
  top: "label"
  data_param {
    source: "/tmp/User1/User1_train_lmdb"
    backend: LMDB
    batch_size: 5
  }
  include: { phase: TRAIN }
}
layers {
  name: "data"
  type: DATA
  top: "data"
  top: "label"
  data_param {
    source: "/tmp/User1/User1_val_lmdb"
    backend: LMDB
    batch_size: 5
  }
  include: { phase: TEST }
}
layers {
  name: "conv1"
  type: CONVOLUTION
  bottom: "data"
  top: "conv1"
  blobs_lr: 1
  blobs_lr: 2
  weight_decay: 1
  weight_decay: 0
  convolution_param {
    num_output: 32
    kernel_size: 11
    stride: 1
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 0
    }
  }
}
layers {
  name: "relu1"
  type: RELU
  bottom: "conv1"
}
layers {
  name: "pool2"
  type: POOLING
  bottom: "conv1"
  top: "pool2"
  pooling_param {
    pool: MAX
    kernel_size: 3
    stride: 2
  }
}
layers {
  name: "norm1"
  type: LRN
  bottom: "pool2"
  top: "norm1"
  lrn_param {
    local_size: 5
    alpha: 0.0001
    beta: 0.75
  }
}
layers {
  name: "conv3"
  type: CONVOLUTION
  bottom: "norm1"
  top: "conv3"
  blobs_lr: 1
  blobs_lr: 2
  weight_decay: 1
  weight_decay: 0
  convolution_param {
    num_output: 16
    kernel_size: 9
    stride: 1
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
      value: 1
    }
  }
}
layers{
  name: "relu2"
  bottom: "conv3"
  top: "conv3"
  type: RELU
}
layers{
  name: "norm2"
  type: LRN
  bottom: "conv3"
  top: "norm2"
  lrn_param{
    local_size: 5
    alpha: 0.0001
    beta: 0.75
  }
}

layers{
  name: "local4"
  type: LOCAL
  bottom: "norm2"
  top: "local4"
  blobs_lr: 1
  blobs_lr: 2
  local_param{
    num_output: 4
    kernel_size: 9
    stride:1
    pad:0
    weight_filler{
      type: "xavier"
    }
    bias_filler{
      type: "constant"
    }
  }
}

layers{
  name: "relu3"
  bottom: "local4"
  top: "local4"
  type: RELU
}

layers{
  name: "local5"
  type: LOCAL
  bottom: "local4"
  top: "local5"
  blobs_lr: 1
  blobs_lr: 2
  local_param{
    num_output: 4
    kernel_size: 7
    stride:2
    pad:0
    weight_filler{
      type: "xavier"
    }
    bias_filler{
      type: "constant"
    }
  }
}

layers{
  name: "relu4"
  type:RELU
  bottom: "local5"
  top: "local5"
}

layers{
  name: "local6"
  type: LOCAL
  bottom: "local5"
  top: "local6"
  blobs_lr: 1
  blobs_lr: 2
  local_param{
    num_output: 4
    kernel_size: 5
    stride:1
    pad:0
    weight_filler{
      type: "xavier"
    }
    bias_filler{
      type: "constant"
    }
  }
}

layers{
  name: "relu5"
  bottom: "local6"
  top: "local6"
  type: RELU
}

layers {
  name: "fc7"
  type: INNER_PRODUCT
  bottom: "local6"
  top: "fc7"
  blobs_lr: 1
  blobs_lr: 2
  weight_decay: 1
  weight_decay: 0
  inner_product_param {
    num_output: 4096
    weight_filler {
      type: "gaussian"
      std: 0.005
    }
    bias_filler {
      type: "constant"
    }
  }
}
layers {
  name: "relu6"
  type: RELU
  bottom: "fc7"
  top: "fc7"
}
layers {
  name: "drop7"
  type: DROPOUT
  bottom: "fc7"
  top: "fc7"
  dropout_param {
    dropout_ratio: 0.5
  }
}
layers {
  name: "fc8"
  type: INNER_PRODUCT
  bottom: "fc7"
  top: "fc8"
  blobs_lr: 1
  blobs_lr: 2
  weight_decay: 1
  weight_decay: 0
  inner_product_param {
    num_output: 6
    weight_filler {
      type: "gaussian"
      std: 0.005
    }
    bias_filler {
      type: "constant"
    }
  }
}
layers {
  name: "accuracy"
  type: ACCURACY
  bottom: "fc8"
  bottom: "label"
  top: "accuracy"
  include: { phase: TEST }
}
layers {
  name: "loss"
  type: SOFTMAX_LOSS
  bottom: "fc8"
  bottom: "label"
  top: "loss"
}

**Comment by jasonustc** [comment]
Has any body comes into the speed issue? I build a model that has 2 conv layers and 2 local layers, and use a K80 GPU, but the training speed is very slow, only 50 epoch in 3 minutes, it seems a bit wired... Here is my prototxt for CIFAR-10
```
name: "CIFAR10_do"
layer {
  name: "cifar"
  type: "Data"
  top: "data"
  top: "label"
  include {
    phase: TRAIN
  }
  transform_param{
	mirror: true
	crop_size: 24
	mean_file: "D:\\Users\\v-xushe\\caffe\\cifar10\\data\\train_cifar10_mean"
  }
  data_param {
    source: "D:\\Users\\v-xushe\\caffe\\cifar10\\data\\cifar10_train_leveldb_2"
    batch_size: 128
#default database type is leveldb
  }
}
layer {
  name: "cifar"
  type: "Data"
  top: "data"
  top: "label"
  include {
    phase: TEST
  }
  data_param {
    source: "D:\\Users\\v-xushe\\caffe\\cifar10\\data\\cifar10_test_leveldb_2"
    batch_size: 100
#default database type is leveldb
  }
  transform_param{
    crop_size: 24
	mean_file: "D:\\Users\\v-xushe\\caffe\\cifar10\\data\\train_cifar10_mean"
  }
}
layer {
  name: "conv1"
  type: "Convolution"
  bottom: "data"
  top: "conv1"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  convolution_param {
    num_output: 64
    pad: 2
    kernel_size: 5
    stride: 1
    weight_filler {
      type: "gaussian"
      std: 0.0001
    }
    bias_filler {
      type: "constant"
    }
  }
}
layer {
  name: "pool1"
  type: "Pooling"
  bottom: "conv1"
  top: "pool1"
  pooling_param {
    pool: MAX
    kernel_size: 3
    stride: 2
  }
}
layer {
  name: "relu1"
  type: "ReLU"
  bottom: "pool1"
  top: "pool1"
}
layer {
  name: "norm1"
  type: "LRN"
  bottom: "pool1"
  top: "norm1"
  lrn_param {
    local_size: 9
    alpha: 0.001
    beta: 0.75
    norm_region: WITHIN_CHANNEL
  }
}
layer {
  name: "conv2"
  type: "Convolution"
  bottom: "norm1"
  top: "conv2"
  param {
    lr_mult: 1
  }
  param {
    lr_mult: 2
  }
  convolution_param {
    num_output: 64
    pad: 2
    kernel_size: 5
    stride: 1
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
    }
	group: 8
  }
}
layer {
  name: "relu2"
  type: "ReLU"
  bottom: "conv2"
  top: "conv2"
}

layer {
  name: "norm2"
  type: "LRN"
  bottom: "conv2"
  top: "norm2"
  lrn_param {
    local_size: 3
    alpha: 0.001
    beta: 0.75
    norm_region: WITHIN_CHANNEL
  }
}

layer {
  name: "pool2"
  type: "Pooling"
  bottom: "norm2"
  top: "pool2"
  pooling_param {
    pool: MAX
    kernel_size: 3
    stride: 2
  }
}


layer {
  name: "lc1"
  type: "Local"
  bottom: "pool2"
  top: "lc1"
  local_param {
    num_output: 64
	pad: 1
	stride: 1
	kernel_size: 3
	
    weight_filler {
      type: "gaussian"
      std: 0.04
    }
    bias_filler {
      type: "constant"
    }
  }
}

layer {
  name: "relu_ip1"
  type: "ReLU"
  bottom: "lc1"
  top: "lc1"
}

layer {
  name: "lc2"
  type: "Local"
  bottom: "lc1"
  top: "lc2"
  local_param {
    num_output: 64
	pad: 1
	stride: 1
	kernel_size: 3
    weight_filler {
      type: "gaussian"
      std: 0.04
    }
    bias_filler {
      type: "constant"
    }
  }
}

layer {
  name: "relu_lc2"
  type: "ReLU"
  bottom: "lc2"
  top: "lc2"
}

layer {
  name: "ip1"
  type: "InnerProduct"
  bottom: "lc2"
  top: "ip1"
  inner_product_param {
    num_output: 128
    weight_filler {
      type: "gaussian"
      std: 0.01
    }
    bias_filler {
      type: "constant"
    }
  }
}

layer {
	name: "relu_ip1"
	type: "ReLU"
	bottom: "ip1"
	top: "ip1"
}

layer{
	name: "drop1"
	type: "Dropout"
	bottom: "ip1"
	top: "drop1"
	dropout_param {
		dropout_ratio: 0.5
#		drop_type: GAUSSIAN
	}
}

layer {
  name: "ip2"
  type: "InnerProduct"
  bottom: "drop1"
  top: "ip2"
  inner_product_param {
    num_output: 10
    weight_filler {
      type: "gaussian"
      std: 0.1
    }
    bias_filler {
      type: "constant"
    }
  }
}


layer {
  name: "accuracy"
  type: "Accuracy"
  bottom: "ip2"
  bottom: "label"
  top: "accuracy"
  include {
    phase: TEST
  }
}
layer {
  name: "loss"
  type: "SoftmaxWithLoss"
  bottom: "ip2"
  bottom: "label"
  top: "loss"
}
```

**Comment by bamos** [comment]
Hi, what's the status on this issue? What does the `ES` tag mean?

I'm interested in merging this into the master branch to use some of the features there.
Is anybody else also interested in this?

I'm happy to help with the merge, but I'm not familiar with Caffe's internals.

-Brandon.

**Comment by swamiviv** [comment]
I want to use the locally connected layers. Can anyone help with a merge ?

-Swami

**Comment by shelhamer** [comment]
Closing since the dev branch is deprecated. Please send PRs to master.

**Comment by kailizhao** [comment]
@jackculpepper  I go though your codes and merged it to the caffe(2015.05 version). I wanna know the difference between your local layer from  the convolution layer in codes. Are the parameters of caffe_cpu_gemm different?
Thanks

**Comment by museless** [comment]
No, I still learning the code and don't modify the code.

**Comment by nitish11** [comment]
@jackculpepper  : I an unable to build this PR.
Getting the following error : 
-- Installing: DeepFace/caffe-b4600f5f84771a3c39d78e31b2d1a3ba75544d9e/build/install/python/caffe/proto/__init__.py
CMake Error at src/caffe/proto/cmake_install.cmake:40 (FILE):
  file INSTALL cannot find
  "DeepFace/caffe-b4600f5f84771a3c39d78e31b2d1a3ba75544d9e/build/src/caffe/proto/caffe_pb2.py".
Call Stack (most recent call first):
  src/caffe/cmake_install.cmake:41 (INCLUDE)
  cmake_install.cmake:42 (INCLUDE)

Any check on this??
make: *** [install] Error 1

**Comment by nitish11** [comment]
I got it resolved, I used the build below and it worked.
https://github.com/BVLC/caffe/tree/96b20185f77d17e81efead841a18aa509f9f7c4f

**Comment by tanfei2007** [comment]
@jackculpepper 
How many overlapping regions  for your local connected layers as mentioned in paper[1] ? 
in your code, it seems that each unit in a feature map have unique filter. 

[1] http://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=6247968

**Comment by DaleSong89** [comment]
It seems that the implemention does not support the local-conv manner you mentioned.@tanfei2007

**Comment by chenxinhua** [comment]
in caffe,  can we stack splitsLayer,cropLayer,convolutionLayer,flattenLayer, concatLayer  to  produce  locally convolution layer？

**Comment by chenxinhua** [comment]
in caffe,  can we stack splitsLayer,cropLayer,convolutionLayer,flattenLayer, concatLayer  to  produce  locally convolution layer？
